### PR TITLE
Fix parsing of BPM and CustomTransferMap from Elegant

### PR DESCRIPTION
**Pull request by Hespe**
## Description

The Elegant `MONI` element is currently not properly mapped to the Cheetah `BPM`. Furthermore, there is a slight issue with how `MONI` with non-zero length are handled.
Similarly the current parser for `CustomTransferMap` is faulty since the `dtype` of the resulting tensor cannot be properly inferred.

## Motivation and Context

- [x] I have raised an issue to propose this change
  - #272 
  - #274 

## Types of changes

This PR is a breaking change, since elegant `MONI` elements are now mapped to `BPM` instead of `Marker`, requiring changes to downstream code.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (_required for a bug fix or a new feature_).
- [ ] I have updated the documentation accordingly.
  
  
- [x] I have reformatted the code and checked that formatting passes (**required**).
- [x] I have have fixed all issues found by `flake8` (**required**).
- [x] I have ensured that all `pytest` tests pass (**required**).
- [ ] I have run `pytest` on a machine with a CUDA GPU and made sure all tests pass (**required**).
- [x] I have checked that the documentation builds (**required**).

Note: We are using a maximum length of 88 characters per line.

### CODE REVIEW

File: tests/test_elegant_conversion.py [hunk 1]
```diff
@@ -32,14 +33,42 @@ def test_fodo():
...
-    assert converted.s1.e1 == correct_lattice.s1.e1
+    assert torch.isclose(converted.q1.length, correct_lattice.q1.length)
+    assert torch.isclose(converted.q1.k1, correct_lattice.q1.k1)
+    assert torch.isclose(converted.q2.length, correct_lattice.q2.length)
+    assert torch.isclose(converted.q2.k1, correct_lattice.q2.k1)
+    for i in range(2):
+        assert torch.isclose(converted.d1[i].length, correct_lattice.d1[i].length)
+    assert torch.isclose(converted.d2.length, correct_lattice.d2.length)
+    assert torch.isclose(converted.s1.length, correct_lattice.s1.length)
+    assert torch.isclose(converted.s1.e1, correct_lattice.s1.e1)
```
Comment (jank324) [inline tests/test_elegant_conversion.py:44]: I feel like this should be no loop and use `torch.allclose`.

File: tests/test_elegant_conversion.py [hunk 2]
```diff
@@ -32,14 +33,42 @@ def test_fodo():
...
+            [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+        ]
+    )
+
+    assert np.isclose(converted.c1.length, 0.7)
+    assert np.isclose(converted.c1.frequency, 1.2e9)
+    assert np.isclose(converted.c1.voltage, 16.175e6)
+    for i in range(2):
+        assert torch.allclose(converted.c1e[i]._transfer_map, correct_map)
```
Comment (jank324) [inline tests/test_elegant_conversion.py:68]:
```suggestion
    correct_transfer_map = torch.tensor(
        [
            [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [0.04, 1.0, 0.003, 0.0, 0.0, 0.0, -0.0027],
            [0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0],
            [0.003, 0.0, -0.04, 1.0, 0.0, 0.0, -0.15],
            [0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0],
            [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0],
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        ]
    )

    assert np.isclose(converted.c1.length, 0.7)
    assert np.isclose(converted.c1.frequency, 1.2e9)
    assert np.isclose(converted.c1.voltage, 16.175e6)
    for i in range(2):
        assert torch.allclose(converted.c1e[i]._transfer_map, correct_transfer_map)
```

File: tests/test_elegant_conversion.py [hunk 3]
```diff
@@ -32,14 +33,42 @@ def test_fodo():
...
+    assert torch.isclose(converted.q2.k1, correct_lattice.q2.k1)
+    for i in range(2):
+        assert torch.isclose(converted.d1[i].length, correct_lattice.d1[i].length)
+    assert torch.isclose(converted.d2.length, correct_lattice.d2.length)
+    assert torch.isclose(converted.s1.length, correct_lattice.s1.length)
+    assert torch.isclose(converted.s1.e1, correct_lattice.s1.e1)
+
+
+def test_cavity_import():
+    """Test importing an accelerating cavity defined in the Elegant file format."""
```
Comment (jank324) [inline tests/test_elegant_conversion.py:48]: I feel like this is a weird instance of cavity import and the docstring (and maybe even the name) of the test should reflect and explain that.

Comment (jank324) [reply tests/test_elegant_conversion.py:48]: On second thoughts, this is really two separate tests, one testing if the cavity was imported correctly (and without throwing an exception) and another doing the save for a custom transfer map, right?

File: tests/test_elegant_conversion.py (hunk 1 above)
Comment (Hespe) [reply tests/test_elegant_conversion.py:44]: I`m not sure how `torch.allclose` would work here, but maybe there is some trick I`m missing. The problem is that we have two elements with name `d1` and therefore two separate `length` tensors, not one multidimensional tensor.

File: tests/test_elegant_conversion.py (hunk 3 above)
Comment (Hespe) [reply tests/test_elegant_conversion.py:48]: I`ve decided to split the test into two separate parts. One for testing the regular cavity import, the second for the custom transfer map.

File: tests/test_elegant_conversion.py (hunk 1 above)
Comment (jank324) [reply tests/test_elegant_conversion.py:44]: Oh I see ... I misunderstood the that line 🙃

**Review by jank324 — APPROVED** [review]

**Pull request merged**